jupyterhub/app.py

```diff
@@ -1975,25 +1975,63 @@ async def init_groups(self):
                 group.users.append(user)
         db.commit()
 
-    async def init_roles(self):
+    async def init_role_creation(self):
         """Load default and predefined roles into the database"""
-        db = self.db
-        # tokens are added separately
-        role_bearers = ['users', 'services', 'groups']
-
-        # load default roles
         self.log.debug('Loading default roles to database')
         default_roles = roles.get_default_roles()
-        for role in default_roles:
-            roles.create_role(db, role)
+        config_role_names = [r['name'] for r in self.load_roles]
+        init_roles = self.load_roles.copy()
+        for role_name in config_role_names:
+            if config_role_names.count(role_name) > 1:
+                raise AttributeError(
+                    f"Role {role_name} multiply defined. Please check the `load_roles` configuration"
+                )
+        for role_spec in default_roles:
+            if role_spec['name'] not in config_role_names:
+                init_roles.append(role_spec)
+        if not orm.Role.find(self.db, name='admin'):
+            self._rbac_upgrade = True
+        self.db.query(orm.Role).delete()
+        app_log.info("Deleting all roles in database")
+        self.db.commit()
+        for role in init_roles:
+            roles.create_role(self.db, role)
 
+    async def init_role_assignment(self):
+        # tokens are added separately
+        role_bearers = ['users', 'services', 'groups']
+        admin_role_objects = ['users', 'services']
+        config_admin_users = set(self.admin_users) | self.authenticator.admin_users
```

Comment:
Don't need a reference to the long-deprecated `self.admin_users` here. It's already copied with a warning into `authenticator.admin_users` by this point.